File: leetcode/microsoft/older/200_Number_Of_Islands.cpp
// 200. Number of Islands
// https : // leetcode.com/problems/number-of-islands/

#include <vector>

using std::vector;

class Solution
{
private:
    // Perform a dfs from a given node
    void dfs(size_t r, size_t c, vector<vector<char>> &grid)
    {
        // Current node is out of bounds
        if (r < 0 || r >= grid.size() || c < 0 || c >= grid[0].size())
            return;

        // Current node is water
        if (grid[r][c] == '0')
            return;

        // Mark the current node as visited
        grid[r][c] = '0';

        // Explore neighbors
        dfs(r - 1, c, grid); // top
        dfs(r, c + 1, grid); // right
        dfs(r + 1, c, grid); // down
        dfs(r, c - 1, grid); // left
    }

public:
    int numIslands(vector<vector<char>> &grid)
    {
        size_t rows { grid.size() }, cols { grid[0].size() };

        int totalIslands { 0 };

        for (size_t r = 0; r < rows; ++r) {
            for (size_t c = 0; c < cols; ++c) {
                if (grid[r][c] == '1') {
                    dfs(r, c, grid);
                    totalIslands += 1;
                }
            }
        }

        return totalIslands;
    }
};

File: leetcode/graphs/200_number_of_islands.py
# 200. Number of Islands
# https://leetcode.com/problems/number-of-islands/

from typing import List

class Solution:
    def numIslands(self, grid: List[List[str]]) -> int:

        rows, cols = len(grid), len(grid[0])
        dirs = [(1, 0), (-1, 0), (0, 1), (0, -1)]

        def explore_island(r: int, c: int) -> None:
            grid[r][c] = '0' # mark tile as visited
            for dr, dc in dirs:
                nr, nc = r + dr, c + dc
                # Check for land in all 4 directions
                if nr >= 0 and nr < rows and nc >= 0 and nc < cols and grid[nr][nc] == '1':
                    explore_island(nr, nc)

        islands_count = 0

        for r in range(rows):
            for c in range(cols):
                if grid[r][c] == '1':
                    explore_island(r, c)
                    islands_count += 1

        # print(islands_count)
        return islands_count 




s = Solution()
print(s.numIslands([
    ["1","1","1","1","0"],
    ["1","1","0","1","0"],
    ["1","1","0","0","0"], 
    ["0","0","0","0","0"]
]))  # 1

print(s.numIslands([
    ["1","1","0","0","0"],
    ["1","1","0","0","0"],
    ["0","0","1","0","0"],
    ["0","0","0","1","1"]
]))  # 3

